### PR TITLE
added logic to create file references and added logging

### PR DESCRIPTION
**Pull request by sherwoodf**
ticket: https://app.clickup.com/t/8698qrqq9

### CODE REVIEW

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/FileReference.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    path_to_search = crate_path / dataset.id / "*"
+    paths = glob.glob(str(path_to_search), recursive=True)
+    return paths
+
+
+def create_api_file_reference(
+    file_path: str, study_uuid: str, dataset_uuid: str, crate_path: pathlib.Path
+) -> list[APIModels.FileReference]:
+    
+    relative_path = pathlib.Path(file_path).relative_to(crate_path).as_posix()
```
Comment (sherwoodf) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/FileReference.py:56]: using as_posix because i'm assuming our system will be posix & we don't want this (and therefore the UUID) to vary depending on whether a windows system was used to ingest the objects.

File: bia_ro_crate/cli.py [hunk 2]
```diff
@@ -78,15 +79,9 @@ def convert(
...
-    #     Path(__file__).parents[0]
-    #     / "model"
-    #     / "example"
-    #     / "S-BIAD1494"
-    #     / "ro-crate-version"
-    # )
 
-    crate = crate_read(crate_path)
+    # Just for validation
+    crate_read(crate_path)
```
Comment (kbab) [inline bia_ro_crate/cli.py:84]: Is the reason for doing this to get a more informative error message than what would be obtained by `process_ro_crate` in the event that the contents are invalid?

File: bia_ro_crate/model/example/S-BIAD1494/ro-crate-version/ro-crate-metadata.json [hunk 3]
```diff
@@ -226,7 +227,7 @@
             "@type": "@id"
         },
         "association": {
-            "@id": "biashapes:association",
+            "@id": "bias:association",
```
Comment (kbab) [inline bia_ro_crate/model/example/S-BIAD1494/ro-crate-version/ro-crate-metadata.json]: is this `bias:` as opposed to `bia:` on purpose?

File: bia_ro_crate/ro_crate_to_bia/crate_reader.py
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/crate_reader.py:1]: should there be a `@TODO` at line 32 of this file (i.e. the definition of the `validate_json` function?

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/AnnotationMethod.py [hunk 4]
```diff
@@ -27,7 +29,7 @@ def create_api_image_acquisition_protocol(
 def convert_annotation_method(
     ro_crate_annotation_method: ROCrateModels.AnnotationMethod,
     crate_objects_by_id: dict[str, ROCrateModel],
-    study_uuid: UUID,
+    study_uuid: str,
```
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/AnnotationMethod.py:30]: Does this function need `crate_objects_by_id`?

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/Dataset.py
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/Dataset.py:1]: Should `convert_image_acquisition_protocol` be changed to `convert_dataset`? Also the function does not use `crate_objects_by_id` parameter.

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/FileReference.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    # TODO: Work out how file URI would be generated.
+
+    file_reference = {
+        "uuid": str(uuid_creation.create_file_reference_uuid(relative_path, study_uuid)),
+        "submission_dataset_uuid": dataset_uuid,
+        "file_path": str(relative_path),
+        "version": 1,
+        "size_in_bytes": pathlib.Path(file_path).stat().st_size,
+        "format": pathlib.Path(file_path).suffix,
```
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/FileReference.py]: This is interesting. I think in the original `bia_shared_models` this property was from biostudies - which distinguished between *file* / *directory* (and I think we also used this in the past to flag files in zip archives).

During biostudies ingest we take this from the `BioStudiesAPIFile` object and its value is usually 'file'. However, EMPIAR ingest populates this with the suffix of the file path (compare value of `format` from example from [biostudies ingest](https://wwwdev.ebi.ac.uk/bioimage-archive/api/v2/file_reference/000012e3-8fb6-4c46-a9cf-2247b76cedf1) with [EMPIAR ingest](https://wwwdev.ebi.ac.uk/bioimage-archive/api/v2/file_reference/00000bbb-9d9f-4a10-8d2e-b33c42541a4d)).

If we decide to go this way it would be useful to know whether to identify special suffixes e.g. (ome.zarr, nii.gz, ome.zarr.zip, zarr.zip, ome.tiff) and whether to standardise suffixes (e.g. TIF vs tif vs tiff vs TIFF) - we have a function for this when creating `images/image_representations`

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/ImageAcquisitionProtocol.py
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/ImageAcquisitionProtocol.py:1]: function does not use `crate_objects_by_id` parameter.

File: bia_ro_crate/ro_crate_to_bia/entity_conversion/Protocol.py
Comment (kbab) [inline bia_ro_crate/ro_crate_to_bia/entity_conversion/Protocol.py:1]: function does not use `crate_objects_by_id` parameter.

File: test/ro_crate_to_bia/output.json
Comment (kbab) [inline test/ro_crate_to_bia/output.json:1]: I skipped this file - I assume that's OK.

File: test/test_ro_crate_to_bia.py [hunk 6]
```diff
@@ -27,4 +27,7 @@ def test_ingest_ro_crate_metadata(tmp_path: Path):
     with open(Path(__file__).parent / "ro_crate_to_bia" / "output.json") as f:
         expected_out = json.loads(f.read())
 
-    assert cli_out == expected_out
+    # Account for different ordering of JSON objects due to file reference order being somewhat arbitrary.
+    assert len(cli_out) == len(expected_out)
+    for json_obj in cli_out:
```
Comment (kbab) [inline test/test_ro_crate_to_bia.py]: Could there be a scenario where two identical objects are created instead of two distinct ones? e.g. If we have `FileReference1` twice instead of `FileReference1` and `FileReference2`. In such a scenario the above will pass the test. However, if the for loop and following assertion explicitly go through all expected objects, the test will fail.

**Review by kbab — COMMENTED** [review]
Left some comments - also not sure if you are currently writing to an API. I think the API versions start from 0 not 1. If this is the case the `version` in the functions converting ROCrate models to API ones needs to change.

File: bia_ro_crate/cli.py (hunk 2 above)
Comment (sherwoodf) [reply bia_ro_crate/cli.py:84]: Yep - This is to check that the crate passes the validation set in (this particular) python library that reads ro-crates.

I actually suspect we might move away from this - mainly because this library makes assumptions that some of the 'SHOULD' things from the spec are followed - and i suspect we don't want to follow all of these. It also makes assumptions about what the @context of the document contains. But i think i want the change away from this form of the ro-crate json to be it's own PR, since it comes with some of these more strategic decisions.

File: test/test_ro_crate_to_bia.py (hunk 6 above)
Comment (sherwoodf) [reply test/test_ro_crate_to_bia.py]: Oh interesting edge case - will update as you suggest

**Review by kbab — APPROVED** [review]
LGTM
